### PR TITLE
Show rooms in common with another user

### PR DESCRIPTION
**Pull request by Half-Shot**
This PR adds the ability to show the rooms you have in common with another user.

As far as I'm aware, there is no design for this yet so it would be good to get some input at some point. This is inspired by Telegram's "groups in common" feature.

shared_rooms_feature

Additionally, this now shows a subset of those rooms (up to 3) in the invite room preview screen:

image

Meta-issue: https://github.com/vector-im/riot-web/issues/14339
JS-SDK PR: https://github.com/matrix-org/matrix-js-sdk/pull/1417
Synapse PR:  https://github.com/matrix-org/synapse/pull/7785

### CODE REVIEW

File: src/components/views/right_panel/UserInfoSharedRooms.tsx [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+                room={room}
+                url={new SpecPermalinkConstructor().forRoom(alias)}
+                inMessage={false}
+                shouldShowPillAvatar={true}
+                isSelected={false}
+            /></a>;
+        }
+
+        return <li key={room.roomId}>
+            <RoomTile
```
Comment (turt2live) [inline src/components/views/right_panel/UserInfoSharedRooms.tsx:126]: just fyi that this component is disappearing in the new room list. Unfortunately the new room tile isn't meant to be used in this way either, so if product does approve this then it'll need to go through design review and get a custom implementation.

**Comment by nadonomy** [comment]
Definitely want this functionality, but need to think about how to expose it in the right panel in more detail so we don't add to the stockpile of design debt. Will look at this in earnest but more like a p2-1 where we have a few p1's in flight atm.

File: src/components/views/right_panel/UserInfoSharedRooms.tsx (hunk 1 above)
Comment (Half-Shot) [reply src/components/views/right_panel/UserInfoSharedRooms.tsx]: This now has a `UserInfoRoomTile`

File: src/components/views/right_panel/UserInfo.js [hunk 2]
```diff
@@ -1312,6 +1313,8 @@ const BasicUserInfo = ({room, member, groupId, devices, isRoomEncrypted}) => {
     const isMe = member.userId === cli.getUserId();
     const canVerify = cryptoEnabled && homeserverSupportsCrossSigning && !userVerified && !isMe;
 
+    const isSharedRoomsFeatureEnabled = SettingsStore.isFeatureEnabled("feature_show_shared_rooms");
```
Comment (tulir) [inline src/components/views/right_panel/UserInfo.js:1316]:
```suggestion
    const isSharedRoomsFeatureEnabled = SettingsStore.getValue("feature_show_shared_rooms");
```

File: src/components/views/rooms/RoomPreviewBar.js [hunk 3]
```diff
@@ -476,13 +479,18 @@ export default createReactClass({
                 secondaryActionLabel = _t("Reject");
                 secondaryActionHandler = this.props.onRejectClick;
 
+                if (SettingsStore.isFeatureEnabled("feature_show_shared_rooms")) {
```
Comment (tulir) [inline src/components/views/rooms/RoomPreviewBar.js:482]:
```suggestion
                if (SettingsStore.getValue("feature_show_shared_rooms")) {
```

File: src/settings/Settings.ts [hunk 4]
```diff
@@ -176,6 +176,12 @@ export const SETTINGS: {[setting: string]: ISetting} = {
         supportedLevels: LEVELS_DEVICE_ONLY_SETTINGS,
         default: false,
     },
+    "feature_show_shared_rooms": {
+        supportedLevels: LEVELS_ACCOUNT_SETTINGS,
```
Comment (tulir) [inline src/settings/Settings.ts:180]: all the other labs settings seem to be on this level:

```suggestion
        supportedLevels: LEVELS_FEATURE,
```

File: src/components/views/elements/UserInfoRoomTile.tsx [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        );
+
+        let ariaLabel = name;
+        let ariaDescribedBy: string;
+
+        const props: Partial<React.ComponentProps<typeof AccessibleTooltipButton>> = {};
+        let Button: React.ComponentType<React.ComponentProps<typeof AccessibleButton>> = AccessibleButton;
+
+        return (
+            <React.Fragment>
```
Comment (resynth1943) [inline src/components/views/elements/UserInfoRoomTile.tsx:101]:
```suggestion
            <>
```

**Comment by nadonomy** [comment]
@Half-Shot is this PR stale? Is it possible to get a test build, or some screenshots of how it looks now post rebrand?

**Comment by Half-Shot** [comment]
PR should still be fine. I think @tulir is running it and he can probably screenshot you?

**Comment by jryans** [comment]
@Half-Shot If you merge develop, Netlify should [add a preview site](https://github.com/vector-im/element-web/blob/develop/docs/pr-previews.md) to this PR, which may be helpful for Product review here.

**Comment by Half-Shot** [comment]
This got broken by the removal of roomListSorter, I'll have to look at this in more depth.

**Comment by Half-Shot** [comment]
@jryans sadly it looks like the preview doesn't show labs by default, which makes them a bit useless :(

**Comment by ShadowJonathan** [comment]
@Half-Shot What's the next step with this? What're we waiting for here?

**Comment by Half-Shot** [comment]
Waiting for Design to get back to me on feedback and possibly a mockup to work against. I don't think there is any dev task for me.

**Comment by Half-Shot** [comment]
N4d has got back to me with feedback, and I shall begin implementing :)

**Comment by nadonomy** [comment]
@Half-Shot just checking in on this as it's still on my list. :) If I'm blocking anything on this it may have slipped through the cracks on my side, so please DM if so.

**Comment by Half-Shot** [comment]
@nadonomy it's still on me to do the changes we talked about :)

**Comment by Half-Shot** [comment]
I've made some changes so that the shared room list now appears in the right panel. There is a lot left to do though:

- [ ] Design on invite / user info panel for the shared users button needs work
- [ ] No way to search / paginate the shared room list atm
- [ ] Dispatching right panel actions from the invite dialogue doesn't work

**Comment by tulir** [comment]
The new panel for shared rooms seems to be missing a back button to go back to the profile view or member list. The x button just closes the whole panel.

**Comment by t3chguy** [comment]
So looks like the API doesn't work on m.org which is quite a downer to this getting any further traction:

image

**Comment by tulir** [comment]
That looks like a misconfiguration on matrix.org, likely related to workers and routing requests

**Comment by t3chguy** [comment]
Blocked on https://github.com/matrix-org/synapse/issues/10339

**Comment by aaronraimist** [comment]
Presumably the PR will fail gracefully if the server doesn't support it?

Why does matrix.org have to support the feature for it to go into Element? m.org doesn't support presence but Element does.

**Comment by t3chguy** [comment]
The PR gracefully fails with a user-presented error, as conditionally showing a section in the user info panel would be confusing

**Comment by MurzNN** [comment]
Instead of "Shared rooms" maybe better to use phrase "Mutual rooms"? Because we already use "Share" therm to share room action, and current "Shared rooms" looks like this person has shared those rooms personally for you.

**Comment by ShadowJonathan** [comment]
This is more a MSC comment, but yes, it has already been noted on the MSC; https://github.com/matrix-org/matrix-doc/pull/2666#discussion_r772331947

After unblocking this on synapse's side, im interested in working on this PR, as half-shot is currently unavailable to finish this work.

**Comment by turt2live** [comment]
This has been open for so long I'm not really sure how to make it go forward at all. 

It's clear we want this feature (per design review), but the implementation is subject to design iteration. The code is also almost certainly bitrotted to a point of no return - even if the merge conflicts can be dealt with, there's non-trivial architectural changes and modern code style to adhere to. 

I think the best thing for this PR is to probably close it and if someone wants to reuse the code then it'll still be here. However, I will note that design work needs to be completed before someone can reasonably pick this up again.

If an issue hasn't been opened, please open one so it can get added to the design backlog.

**Pull request closed without merging**